nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs

```diff
@@ -0,0 +1,427 @@
+use std::net::IpAddr;
+use std::sync::Arc;
+use std::time::Duration;
+
+use ipnetwork::{IpNetwork, Ipv4Network};
+use tokio::task::JoinSet;
+use tokio_util::sync::CancellationToken;
+
+use nym_authenticator_client::AuthClient;
+use nym_gateway_directory::{
+    AuthAddresses, EntryPoint, ExitPoint, Gateway, GatewayClient, Recipient,
+};
+use nym_sdk::UserAgent;
+use nym_task::TaskManager;
+use nym_wg_gateway_client::{GatewayData, WgGatewayClient};
+use nym_wg_go::{PrivateKey, PublicKey};
+
+use crate::mixnet::SharedMixnetClient;
+use crate::platform::VPNConfig;
+use crate::{bandwidth_controller::BandwidthController, GenericNymVpnConfig};
+use crate::{GatewayDirectoryError, MixnetClientConfig};
+
+use super::ios::tun_provider::OSTunProvider;
+use super::{
+    two_hop_tunnel::TwoHopTunnel,
+    wg_config::{WgInterface, WgNodeConfig, WgPeer},
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("mixnet timed out during startup")]
+    StartMixnetTimeout,
+
+    #[error("failed to start mixnet client: {0}")]
+    StartMixnetClient(#[source] Box<dyn std::error::Error + Send>),
+
+    #[error("gateway directory error: {0}")]
+    GatewayDirectory(#[from] GatewayDirectoryError),
+
+    #[error("authenticator address not found")]
+    AuthenticatorAddressNotFound,
+
+    #[error("not enough bandwidth")]
+    NotEnoughBandwidth,
+
+    #[error("wirewireguardurad authentication is not possible due to one of the gateways not running the authenticator process: {0}")]
+    AuthenticationNotPossible(String),
+
+    #[error("wireguard gateway failure: {0}")]
+    WgGatewayClientFailure(#[from] nym_wg_gateway_client::Error),
+
+    #[error("failed to run two hop tunnel: {0}")]
+    Tunnel(#[source] super::Error),
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+const MIXNET_CLIENT_STARTUP_TIMEOUT_SECS: Duration = Duration::from_secs(30);
+const TASK_MANAGER_SHUTDOWN_TIMER_SECS: u64 = 10;
+
+pub struct WgTunnelRunner {
+    gateway_directory_client: GatewayClient,
+    task_manager: TaskManager,
+    generic_config: GenericNymVpnConfig,
+    #[cfg(target_os = "ios")]
+    tun_provider: Arc<dyn OSTunProvider>,
+    shutdown_token: CancellationToken,
+}
+
+impl WgTunnelRunner {
+    pub fn new(config: VPNConfig, shutdown_token: CancellationToken) -> Result<Self> {
+        let user_agent = UserAgent::from(nym_bin_common::bin_info_local_vergen!());
+        tracing::info!("User agent: {user_agent}");
+
+        let generic_config = GenericNymVpnConfig {
+            mixnet_client_config: MixnetClientConfig {
```

Comment:
Unrelated side comment: we should implement `Default` for `MixnetClientConfig` in the future